pkg/provisioner/ironic/power_test.go

```diff
@@ -295,7 +298,52 @@ func TestPowerOff(t *testing.T) {
 			if tc.expectedErrorResult {
 				assert.Contains(t, result.ErrorMessage, "hard power off failed")
 			}
-
+			if tc.expectedReason != "" {
+				assert.Len(t, eventReasons, 1)
+				assert.Contains(t, eventReasons, tc.expectedReason)
+			} else {
+				assert.Empty(t, eventReasons)
+			}
 		})
 	}
 }
+
+func TestSoftPowerOffFallback(t *testing.T) {
+	nodeUUID := "33ce8659-7400-4c68-9535-d10766f07a58"
+	node := nodes.Node{
+		PowerState: powerOn,
+		UUID:       nodeUUID,
+	}
+	ironic := testserver.NewIronic(t).Ready().Node(node).WithNodeStatesPowerUpdate(nodeUUID, http.StatusBadRequest)
+	ironic.Start()
+	defer ironic.Stop()
+	inspector := testserver.NewInspector(t).Ready()
+	inspector.Start()
+	defer inspector.Stop()
+
+	host := makeHost()
+	host.Status.Provisioning.ID = nodeUUID
+	var eventReasons []string
+	publisher := func(reason, message string) {
+		eventReasons = append(eventReasons, message)
+	}
+	auth := clients.AuthConfig{Type: clients.NoAuth}
+	prov, err := newProvisionerWithSettings(host, bmc.Credentials{}, publisher,
+		ironic.Endpoint(), auth, inspector.Endpoint(), auth,
+	)
+	if err != nil {
+		t.Fatalf("could not create provisioner: %s", err)
+	}
+
+	_, err = prov.PowerOff(metal3v1alpha1.RebootModeSoft, false)
```

Comment:
Unfortunately the current mock server does not allow to specify multiple, different responses for the same API call, so I don't think there is an easy way currently to test such scenario without enhancing properly the mock (right now, when a response call is configured, like here in `.WithNodeStatesPowerUpdate(nodeUUID, http.StatusBadRequest)` for example, the mock server will _always_ provide the same answer for all the requests).
I think we could discuss this change in a separate PR as it may be useful also for other tests.

Reply:
Yeah, I agree that would be a helpful feature to have in the mocks. If that were possible then this could have been implemented in TestPowerOff (and the regression would not have happened) without having to write this separate TestSoftPowerOffFallback test with its own logic.